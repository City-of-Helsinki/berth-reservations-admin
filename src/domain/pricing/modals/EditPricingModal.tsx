import React, { FunctionComponent } from 'react';
import { useTranslation } from 'react-i18next';
import { Form, Formik, FormikErrors } from 'formik';
import { Button } from 'hds-react/lib';

import Modal from '../../../common/modal/Modal';
import {
  AdditionalService,
  BerthPrice,
  HarborService,
  WinterStoragePrice,
} from '../PricingPage';
import BerthsPricingFields from './BerthsPricingFields';
import WinterStoragePricingFields from './WinterStoragePricingFields';
import HarborServicesPricingFields from './HarborServicesPricingFields';
import AdditionalServicesFields from './AdditionalServicesFields';
import styles from './modals.module.scss';

export enum EDIT_PRICING_FORM_TYPES {
  BERTHS = 'BERTHS',
  HARBOR_SERVICES = 'HARBOR_SERVICES',
  WINTER_STORAGE = 'WINTER_STORAGE',
  ADDITIONAL_SERVICES = 'ADDITIONAL_SERVICES',
}

export interface EditPricingModalProps {
  onSubmit: () => void;
  isOpen: boolean;
  closeModal: () => void;
  initialValues:
    | BerthPrice
    | WinterStoragePrice
    | HarborService
    | AdditionalService;
  formType: EDIT_PRICING_FORM_TYPES;
}

const EditPricingModal: FunctionComponent<EditPricingModalProps> = ({
  isOpen,
  closeModal,
  formType,
  initialValues,
}) => {
  const [isFilling, setFormIsFilling] = React.useState(false);
  const { t } = useTranslation();

  const validate = () => {
    if (!isFilling) {
      setFormIsFilling(true);
    }
    const errors: FormikErrors<
      BerthPrice | WinterStoragePrice | HarborService | AdditionalService
    > = {};
    return errors;
  };

  const renderFormFields = (formType: EDIT_PRICING_FORM_TYPES) => {
    switch (formType) {
      case EDIT_PRICING_FORM_TYPES.BERTHS:
        return <BerthsPricingFields />;
      case EDIT_PRICING_FORM_TYPES.WINTER_STORAGE:
        return <WinterStoragePricingFields />;
      case EDIT_PRICING_FORM_TYPES.HARBOR_SERVICES:
        return <HarborServicesPricingFields />;
      case EDIT_PRICING_FORM_TYPES.ADDITIONAL_SERVICES:
        return <AdditionalServicesFields />;
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      label={t('pricing.editModalHeading').toUpperCase()}
      toggleModal={closeModal}
    >
      <Formik
        initialValues={initialValues}
        onSubmit={values => console.log('Form submit:', values)}
        validate={validate}
      >
        {({ isSubmitting }: { isSubmitting: boolean }) => (
          <Form>
            {renderFormFields(formType)}
            <div className={styles.buttonRow}>
              <Button color="supplementary" onClick={() => closeModal()}>
                {t('common.cancel')}
              </Button>
              <Button
                type="submit"
                className={styles.alignRight}
                disabled={isSubmitting}
              >
                {t('common.save')}
              </Button>
            </div>
          </Form>
        )}
      </Formik>
    </Modal>
  );
};

export default EditPricingModal;
