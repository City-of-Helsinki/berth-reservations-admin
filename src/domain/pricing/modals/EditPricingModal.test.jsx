import React from 'react';
import { mount } from 'enzyme';
import { Form } from 'formik';

import EditPricingModal, { EDIT_PRICING_FORM_TYPES } from './EditPricingModal';

const berthsData = {
  id: '1',
  width: 2,
  privateCustomer: 116,
  company: 236,
  period: 'season',
};

const winterStorageData = {
  id: '1',
  area: 'Kaisaniemi',
  privateCustomer: 8.5,
  company: 17,
  period: 'season',
};

const harborServicesData = {
  id: '1',
  service: 'mooring',
  price: 28,
  period: 'season',
};

const additionalServicesData = {
  id: '1',
  service: 'trawlerSummerStorage',
  price: 24,
  tax: 24,
  period: 'season',
};

const mockProps = {
  onSubmit: jest.fn(),
  closeModal: jest.fn(),
  isOpen: true,
};

describe('EditPricingModal', () => {
  const getWrapper = (props = {}) =>
    mount(
      <div>
        <EditPricingModal {...mockProps} {...props} />
      </div>
    );

  it('renders normally with berths data', () => {
    const wrapper = getWrapper({
      initialValues: berthsData,
      formType: EDIT_PRICING_FORM_TYPES.BERTHS,
    });
    expect(wrapper.find(Form).html()).toMatchSnapshot();
  });

  it('renders normally with winter storage data', () => {
    const wrapper = getWrapper({
      initialValues: winterStorageData,
      formType: EDIT_PRICING_FORM_TYPES.WINTER_STORAGE,
    });
    expect(wrapper.find(Form).html()).toMatchSnapshot();
  });

  it('renders normally with harbor services data', () => {
    const wrapper = getWrapper({
      initialValues: harborServicesData,
      formType: EDIT_PRICING_FORM_TYPES.HARBOR_SERVICES,
    });
    expect(wrapper.find(Form).html()).toMatchSnapshot();
  });

  it('renders normally with additional services data', () => {
    const wrapper = getWrapper({
      initialValues: additionalServicesData,
      formType: EDIT_PRICING_FORM_TYPES.ADDITIONAL_SERVICES,
    });
    expect(wrapper.find(Form).html()).toMatchSnapshot();
  });
});
