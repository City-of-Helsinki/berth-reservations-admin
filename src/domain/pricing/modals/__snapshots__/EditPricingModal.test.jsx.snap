// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditPricingModal renders normally with additional services data 1`] = `"<form action=\\"#\\"><div class=\\"row\\"><div class=\\"formTitle\\"><div class=\\"formTitleLabel\\">Tietue</div><div class=\\"formTitleValue\\">Lisäpalvelut</div></div></div><hr><div class=\\"grid cols1 row\\"><label><span class=\\"labelText\\">Palvelu</span><select name=\\"service\\" id=\\"service\\" class=\\"select\\"><option value=\\"trawlerSummerStorage\\">Trailerin kesäsäilytys</option><option value=\\"parkingPermit\\">Pysäköintilupa</option><option value=\\"dinghyPlace\\">Jollapaikka</option></select></label></div><div class=\\"grid cols2 row\\"><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"price\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Hinta (€)</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"price\\" placeholder=\\"\\" type=\\"text\\" value=\\"24\\"></div></div><label><span class=\\"labelText\\">ALV</span><select name=\\"tax\\" id=\\"tax\\" class=\\"select\\"><option value=\\"24\\">24&nbsp;%</option></select></label></div><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Aika</span><select name=\\"period\\" id=\\"period\\" class=\\"select\\"><option value=\\"season\\">Kausi</option><option value=\\"month\\">1 kk</option><option value=\\"year\\">Vuosi</option></select></label></div><div class=\\"buttonRow\\"><button type=\\"button\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3Epn-oMAAIeEFU9fA8zgCr\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Peruuta</span></button><button type=\\"submit\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3bo_TEhSbadoDUbxUTgx4J alignRight\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Tallenna</span></button></div></form>"`;

exports[`EditPricingModal renders normally with berths data 1`] = `"<form action=\\"#\\"><div class=\\"row\\"><div class=\\"formTitle\\"><div class=\\"formTitleLabel\\">Tietue</div><div class=\\"formTitleValue\\">Venepaikka</div></div></div><hr><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Leveys</span><select name=\\"width\\" id=\\"width\\" class=\\"select\\"><option value=\\"2\\">2,0 m</option><option value=\\"2.5\\">2,5 m</option><option value=\\"2.75\\">2,75 m</option><option value=\\"3\\">3,0 m</option><option value=\\"4\\">4,0 m</option><option value=\\"5\\">5,0 m</option><option value=\\"5.5\\">5,5 m</option><option value=\\"6\\">6,0 m</option><option value=\\"7\\">7,0 m</option></select></label></div><div class=\\"grid cols2 row\\"><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"privateCustomer\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Yksityinen (€)</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"privateCustomer\\" placeholder=\\"\\" type=\\"text\\" value=\\"116\\"></div></div><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"company\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Yritys (€)</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"company\\" placeholder=\\"\\" type=\\"text\\" value=\\"236\\"></div></div></div><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Aika</span><select name=\\"period\\" id=\\"period\\" class=\\"select\\"><option value=\\"season\\">Kausi</option><option value=\\"month\\">1 kk</option><option value=\\"year\\">Vuosi</option></select></label></div><div class=\\"buttonRow\\"><button type=\\"button\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3Epn-oMAAIeEFU9fA8zgCr\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Peruuta</span></button><button type=\\"submit\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3bo_TEhSbadoDUbxUTgx4J alignRight\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Tallenna</span></button></div></form>"`;

exports[`EditPricingModal renders normally with harbor services data 1`] = `"<form action=\\"#\\"><div class=\\"row\\"><div class=\\"formTitle\\"><div class=\\"formTitleLabel\\">Tietue</div><div class=\\"formTitleValue\\">Satamapalvelut</div></div></div><hr><div class=\\"grid cols1 row\\"><label><span class=\\"labelText\\">Palvelu</span><select id=\\"service\\" class=\\"select\\"><option value=\\"mooring\\">Kiinnitys</option><option value=\\"electricity\\">Sähkö</option><option value=\\"water\\">Vesi</option><option value=\\"wasteCollection\\">Jätehuolto</option><option value=\\"gate\\">Portti</option><option value=\\"lighting\\">Valaistus</option></select></label></div><div class=\\"grid cols2 row\\"><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"price\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Hinta</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"price\\" placeholder=\\"\\" type=\\"text\\" value=\\"28\\"></div></div><label><span class=\\"labelText\\">Yksikkö</span><select name=\\"unit\\" id=\\"unit\\" class=\\"select\\"><option value=\\"%\\">%</option><option value=\\"€\\">€</option></select></label></div><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Aika</span><select name=\\"period\\" id=\\"period\\" class=\\"select\\"><option value=\\"season\\">Kausi</option><option value=\\"month\\">1 kk</option><option value=\\"year\\">Vuosi</option></select></label></div><div class=\\"buttonRow\\"><button type=\\"button\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3Epn-oMAAIeEFU9fA8zgCr\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Peruuta</span></button><button type=\\"submit\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3bo_TEhSbadoDUbxUTgx4J alignRight\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Tallenna</span></button></div></form>"`;

exports[`EditPricingModal renders normally with winter storage data 1`] = `"<form action=\\"#\\"><div class=\\"row\\"><div class=\\"formTitle\\"><div class=\\"formTitleLabel\\">Tietue</div><div class=\\"formTitleValue\\">Talvisäilytyspaikka</div></div></div><hr><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Alue</span><select name=\\"area\\" id=\\"area\\" class=\\"select\\"><option value=\\"Kaisaniemi\\">Kaisaniemi</option></select></label></div><div class=\\"grid cols2 row\\"><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"privateCustomer\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Yksityinen (€)</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"privateCustomer\\" placeholder=\\"\\" type=\\"text\\" value=\\"8.5\\"></div></div><div class=\\"_1cealf9APZDUwwgivaltwb\\"><label for=\\"company\\" class=\\"_1NkZD_iEqeTFRGBnfzQ4zX _195GUCoLSCjwq2j1VtKb93 _2tInY8Y3RXDANhP6_8Ipb_ \\">Yritys (€)</label><div class=\\"_1ru3hdCZ3qhTQ_EVvRwBIh\\"><input class=\\"_35GDyiOFLPVMWYFM4_BMWV CQwbnf7hlQrao8Za4G1Wo\\" id=\\"company\\" placeholder=\\"\\" type=\\"text\\" value=\\"17\\"></div></div></div><div class=\\"grid cols2 row\\"><label><span class=\\"labelText\\">Aika</span><select id=\\"period\\" class=\\"select\\"><option value=\\"season\\">Kausi</option><option value=\\"month\\">1 kk</option><option value=\\"year\\">Vuosi</option></select></label></div><div class=\\"buttonRow\\"><button type=\\"button\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3Epn-oMAAIeEFU9fA8zgCr\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Peruuta</span></button><button type=\\"submit\\" class=\\"_3wgeeaHxiHujNcKzeML5j4 _3aSBFzRPhZcNdul13_gERo _3bo_TEhSbadoDUbxUTgx4J alignRight\\"><span class=\\"_2zMLCVF6mmvV2UqfaBF5K\\">Tallenna</span></button></div></form>"`;
